### PR TITLE
Add API to mount a snapshot as a searchable index

### PR DESCRIPTION
**Pull request by dakrone**
This commit adds an API for mounting a snapshot as a searchable index.

Internally, a request is made to the repository to fetch metadata information necessary from the
snapshot in question, which is then passed to the regular snapshot restore mechanism with some
specialized settings.

Assume you have an existing snapshot repository and snapshot, the API looks like:

```json
POST /my-new-index/_snapshot/mount?wait_for_completion=true
{
  "repository": "my-s3-repo",
  "snapshot": "production-2019-12-11-aoeufao",
  "snapshot_index": "important-data",
  "index_settings": {
    "index.number_of_replicas": 2
  },
  "ignore_index_settings": ["index.refresh_interval"]
}
```

While the request may appear similar to a snapshot restore request, it has slight differences. The
`repository`, `snapshot`, and `snapshot_index` fields are required (all other fields are optional).

The response is identical to the response when restoring a snapshot. In the above example the newly
restored/mounted index will be called `my-new-index`.

Relates to #50999

### CODE REVIEW

**Comment by dakrone** [comment]
Note for reviewers: this adds the API, but I tried to make as minimal changes as possible to the rest of the searchable snapshot infrastructure. This does not remove the "searchable" repository type, and only changes a few assertions made about the repository type where necessary to have mounting work with no issues.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRepository.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRepository.java]: This is "disrupting" the usual lifecycle of `Repository`. On the other hand, `SearchableSnapshotRepository` was useful when the searchable repository was explicitly registered as a delegating repository. I don't think it's needed anymore: the  `DirectoryFactory` logic can be moved back to the plugin class, and instead of passing directly a `BlobContainer` instance to the `SearchableSnapshotDirectory` it could use the `RepositoriesService` to retrieve it on demande. This way we get rid of a class and we take care of reloading the `BlobStoreRepository` instance that could have changed.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotAction.java]: Should we make this uniform we other searchable snapshots related actions?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: This doc needs to be updated

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: Should we use  ̀indices` or `index`?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Nit: extra line

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: This throws a `RepositoryMissingException` which must be caught and passed to the `listener.onFailure()` method

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: What are these exactly?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: I'm wondering if we need these verification - maybe we could just let the Restore action do it for us? (and configure appropriate indices options like ignore_unavailable: false)

**Review by tlrx — COMMENTED** [review]
Thanks@dakrone, and thanks for having minimized the changes. I made a first bunch of comments (I didn't look at the tests yet).

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotAction.java
Comment (dakrone) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotAction.java]: I think it's either conform to the searchable snapshot actions, or treat it more like the snapshot action. Here I decided to treat it like a "snapshot" action (a "mount" action) so people could grant `cluster:admin/snapshot/*` and be able to mount searchable snapshots (since it's essentially a specialized restore)

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (dakrone) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: I believe you can override repository-specific settings on a restore (things like speed), these are passed through to the settings that `RestoreSnapshotRequest` takes

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (dakrone) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: If we don't have these then we can't fetch the `IndexId`, so essentially we'd be sending a half-broken restore (because it would be missing crucial index settings required for searchable snapshots to work) in order to push the error elsewhere, I'd rather do the verification as soon as possible so we don't pass around half-broken requests (and do one fewer internal transport requests)

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (dakrone) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: `index` I think. We can't actually support restoring multiple indices currently, as there's no way to specify index-specific settings when restoring multiple indices (ie, restore `foo`, `bar`, and `baz`, but have different `index_settings` for each of them).

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRepository.java
Comment (dakrone) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRepository.java]: I would leave that up to you and David. I agree this is not the "right" way to do it since there is no longer a need for a `SearchableSnapshotRepository`, I agree that the directory logic can be moved elsewhere when the `SearchableSnapshotRepository` repository type is removed. For now, I was trying to make as small a change as possible to code outside of the scope of the actual request/response/transport.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: I'd prefer we did this TODO in this PR, since I think it would mean we could drop the need for validation and make all the required fields into constructor args (and final) instead.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRepository.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotRepository.java]: I'm with Tanguy here, I think it would be good to remove `SearchableSnapshotRepository` in this PR.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/rest/RestMountSearchableSnapshotAction.java
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/rest/RestMountSearchableSnapshotAction.java]: I think just `POST` is sufficient, in line with the restore API. We're not creating a resource at this URI as `PUT`  would imply.

**Review by DaveCTurner — COMMENTED** [review]
Thanks @dakrone, I left a few small comments too.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: It looks like these settings are unused. I have removed them from the API here, and in the  future we can remove them from the usual restore process too.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: I think throwing is fine, we're inside an `ActionRunnable`?

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/bwc
@elasticmachine please run elasticsearch-ci/default-distro

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: Maybe pass the RestRequest directly to the ObjectParser's context when applying the parsing logic?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: Generating a `XContent` from a request object is useful for the `RestHighLevelClient` only (we don't really need it for testing the server logic) so I think this code should be moved to the `rest-high-level` project in a different PR. We'll have to upgrade the `RestHighLevelClient` to support other searchable snapshot APIs anyway and we can do it later.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: Maybe:
```suggestion
        return "mount snapshot [" + repositoryName + ":" + snapshotName + ":" + snapshotIndexName + "] as index [" + mountedIndexName + "]";
```

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Nit: I think this could be plain Javadoc at the class level (for better visibility)

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: I think this deserves a tests

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Nit: maybe `buildIndexSettings()` ?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Good point @DaveCTurner, but I still think that mixing listener and throwing exception is confusing

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Yes, sorry for the noise. I've been too quick at commenting here.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: This is a good observation. I don't have a strong opinion about it as the `Directory` implementation ensures (from index settings) that we are restoring the snapshot id/index id discovered here; but I agree that a different snapshot and/or index could be resolved later by the transport restore action itself (and if that index has a different number of shards, boom). I see this as a low risk that could maybe be worked around by adding snapshot it and index id to the restore request and check them (if present) during restore?

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequestTests.java
Comment (tlrx) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequestTests.java]: I don't think we should test the request parsing logic here (we have `AbstractRequestTestCase` for this in the client project)

**Review by tlrx — COMMENTED** [review]
This looks good, thanks for picking this work David. I've left some comments, but nothing to worry about.

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequestTests.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequestTests.java]: Ok, removed in 4a7588f2959f7a43e98061d788e9ff8225441cfa

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: ++ done in 85ba30188ae38d8dffd9ea2825e501ca12a733e4

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Sure, done in 6d2e48ceffa25c484fb917cccf1ea928655e8031.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: ++ done in cca756ecba0db5cc514e13568c9413681264254e.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: Ok, I've removed this in 4a7588f2959f7a43e98061d788e9ff8225441cfa and added a note to #50999 to revisit client support later.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: Sure, done in dbe178f8d913d3125b3601acf84a34c52ac952ed.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/MountSearchableSnapshotRequest.java]: They can't be `null` by construction (we use `Objects.requireNonNull` in the constructor).

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: I'm not sure how to write a test of this that isn't vacuous. Any suggestions?

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Yes, it'd involve adding extra data to the restore so that we can re-validate them later. I think we need to do this TBH, or else we have no guarantees that the index metadata we restore corresponds with the snapshot.

I expanded the comment in 414e830129249bff43ccec131543f939c3a1b8b7 but we can address this later.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: I prefer to throw. The `try/catch` is unnecessary and meant I felt compelled to investigate whether the framework was missing something.

**Comment by DaveCTurner** [comment]
Everything addressed @tlrx, I think, this is ready for another look.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: Okay. I've checked other places and it seems it also throws in some cases.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: :+1:

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: I looked at it and I don't see how to easily craft a good test for this. Maybe just ignore my review comment, as you already commented correctly the usage of this thread pool.

Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java]: (and sorry for the noise)

**Review by tlrx — APPROVED** [review]
LGTM - if CI is happy, I'll be happy

**Comment by DaveCTurner** [comment]
Thanks @dakrone and @tlrx 🎉

**Pull request merged**
